Explicitly treat ObjectDescription._doc_field_type_map as an instance variable

ViktorHaag opened the pull request:
Subject: Aims to address [issue #6478](https://github.com/sphinx-doc/sphinx/issues/6478)

### Feature or Bugfix

- Bugfix

### Purpose
Explicitly treat `ObjectDescription._doc_field_type_map` population as an instance variable so that we can be sure that Domains with objects that define their own field types will render properly on the output.

### Detail

- Left `ObjectDescription._doc_field_type_map` declaration alone and in place in
  case there are other factors at play that require it declared on the class.

- Ensured that `ObjectDescription.get_field_type_map()` made sure to populate the instance scoped version of `_doc_field_type_map` and not the class scoped version of the variable. 

### Relates
- #6478

ViktorHaag committed:
Explicitly treat ObjectDescription._doc_field_type_map as an instance variable

- Aims to address [issue #6478](https://github.com/sphinx-doc/sphinx/issues/6478)

- Left ObjectDescription._doc_field_type_map declaration alone and in place in
  case there are other factors at play that require it declared on the class.

- In ObjectDescription constructor, explicitly set _doc_field_type_map's value
  as an instance variable.

- This presumably needs to be built with every instantiation of an
  ObjectDescription or inheriting class, so we're not attempting to "define it
  once on a class and then use it"... that approach seems much harder to reason
  about and get correctly done (as was maybe demonstrated by the refactoring
  that lead to this problem in the linked-to issue).

- Runs through the test suite clean locally except for an unrelated single test
  failure on an image_converter test, which also failed on the base branch
  before my changes here.

ViktorHaag committed:
type declaration for new internal method

ViktorHaag committed:
mypy type annotations for methods added

- type annotation for _process_type_map was incorrect; needed to document the
  positional argument for the method

- __init__ method for ObjectDescription missing type annotation, so added

tk0miya commented on sphinx/directives/__init__.py:
```diff
@@ -70,18 +70,25 @@ class ObjectDescription(SphinxDirective):
     def get_field_type_map(self):
         # type: () -> Dict[str, Tuple[Field, bool]]
-        if self._doc_field_type_map == {}:
-            for field in self.doc_field_types:
```
Could you override `self._doc_field_type_map` here please? For some reasons, I'd not like to add a new method.

ViktorHaag replied:
Hey @tk0miya -- I'm sorry, I'm unsure what you mean here... do you mean to move the code I'd put into `_process_type_map` into the constructor? Or do you mean that you'd prefer for `get_field_type_map` to always rebuild the type map?

ViktorHaag replied:
For example, this suggested change seems lighter and more in line with the existing code, but adds an optional keyword argument to `get_field_type_map`:
``` python
    # Warning: this might be removed in future version. Don't touch this from extensions.
    _doc_field_type_map = {}  # type: Dict[str, Tuple[Field, bool]]

    def __init__(self, *args, **kwargs):
        # type: (*Any, **Any) -> None
        SphinxDirective.__init__(self, *args, **kwargs)
        self._doc_field_type_map = self.get_field_type_map(force_rebuild=True)

    def get_field_type_map(self, force_rebuild=False):
        # type: (bool) -> Dict[str, Tuple[Field, bool]]
        if force_rebuild or self._doc_field_type_map == {}:
            for field in self.doc_field_types:
                for name in field.names:
                    self._doc_field_type_map[name] = (field, False)
                if field.is_typed:
                    typed_field = cast(TypedField, field)
                    for name in typed_field.typenames:
                        self._doc_field_type_map[name] = (field, True)
        return self._doc_field_type_map

```

tk0miya replied:
sorry for poor comment. I meant this code.
```
    def get_field_type_map(self) -> Dict[str, Tuple[Field, bool]]:
        if self._doc_field_type_map == {}:
            self._doc_field_type_map = {}    # OVERRIDE the class variable by instance's
            for field in self.doc_field_types:
                for name in field.names:
                    self._doc_field_type_map[name] = (field, False)

                if field.is_typed:
                    typed_field = cast(TypedField, field)
                    for name in typed_field.typenames:
                        self._doc_field_type_map[name] = (field, True)

        return self._doc_field_type_map
```

ViktorHaag replied:
Oh! I see -- yes, that is much simpler. I have pushed a new commit to change as you suggest.

tk0miya reviewed (changes requested):
LGTM with nits.

ViktorHaag committed:
Simplify to ensure we're lazily setting the instance field-type-map variable

- We can more simply have ObjectDescription.get_field_type_map() make sure that
  it's going to lazily set the value of an instance variable, not the class
  variable, when it populates the field type map.

tk0miya commented on sphinx/directives/__init__.py:
```diff
@@ -73,15 +73,14 @@ class ObjectDescription(SphinxDirective):
             for field in self.doc_field_types:
                 for name in field.names:
                     self._doc_field_type_map[name] = (field, False)
-
```
This PR removes unrelated empty lines. Could you fix this? Then I'll merge this soon.

ViktorHaag replied:
OK

ViktorHaag committed:
repair white space

tk0miya reviewed (approved):
Thank you for quick work!

tk0miya commented:
Merged. Thank you for your work!
